src/lightning/data/cache/writer.py

```diff
@@ -87,15 +105,20 @@ def filled(self) -> bool:
         files = os.listdir(self._cache_dir)
         index_files = [f for f in files if f.endswith(_INDEX_FILENAME)]
         worker_end = _WorkerEnv.detect()
-        self._is_done = len(index_files) == self._distributed_env.world_size * worker_end.world_size
+        world_size = os.getenv("DATA_OPTIMIZER_WORLD_SIZE", None)
+        self._is_done = len(index_files) == (world_size or self._distributed_env.world_size * worker_end.world_size)
         return self._is_done
 
     @property
     def rank(self) -> int:
         """Returns the rank of the writer."""
         if self._rank is None:
-            self._worker_env = _WorkerEnv.detect()
-            self._rank = self._distributed_env.global_rank * self._worker_env.world_size + self._worker_env.rank
+            rank = os.getenv("DATA_OPTIMIZER_GLOBAL_RANK", None)
```

Comment:
why do we need to manage this through a global variable? If we can do it through the state, we should avoid the env variable.

Reply:
Because the cache can be used within a DataLoader worker and within the Data Optimizer and the flows are completely different.